src/ParallelAlgorithms/Amr/Actions/CreateChild.hpp

```diff
@@ -0,0 +1,78 @@
+
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <charm++.h>
+#include <iterator>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "Parallel/Callback.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Phase.hpp"
+#include "ParallelAlgorithms/Amr/Actions/SendDataToChildren.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+
+/// \cond
+namespace db {
+template <typename>
+class DataBox;
+}  // namespace db
+/// \endcond
+
+namespace amr::Actions {
+/// \brief Creates a new element in an ArrayAlgorithm whose id is `child_id`
+///
+/// \details This action is meant to be initially invoked by
+/// amr::Actions::AdjustDomain on the amr::Component.  This action inserts a new
+/// element with id `child_id` in the array referenced by `element_proxy`.  A
+/// Parallel::SimpleActionCallback is passed to the constructor of the new
+/// DistributedObject.  If `child_id` is the last element of `children_ids`, the
+/// Parallel::SimpleActionCallback will invoke amr::Actions::SendDataToChildren
+/// on the element with id `parent_id`.  Otherwise, it will invoke
+/// amr::Actions::CreateChild passing the next element of `children_ids`.
+///
+/// This action does not modify anything in the DataBox
+struct CreateChild {
+  template <typename ParallelComponent, typename DbTagList,
+            typename Metavariables, typename ElementProxy>
+  static void apply(
+      db::DataBox<DbTagList>& /*box*/,
+      Parallel::GlobalCache<Metavariables>& cache, const int /*array_index*/,
+      ElementProxy element_proxy,
+      ElementId<Metavariables::volume_dim> parent_id,
+      ElementId<Metavariables::volume_dim> child_id,
```

Comment:
[optional] The code might be cleaner if you took an index into the `children_ids` array instead of `child_id`.  You could avoid the iterator manipulation, at least.